### PR TITLE
Handle forking with connection pools more safely

### PR DESCRIPTION
**Pull request by jmo-qap**
https://www.psycopg.org/docs/usage.html#thread-and-process-safety reminds you that you shouldn't use libpq connections made in the parent in a fork. However, sometimes one doesn't control the forking and then one spends a lot of time debugging ... :-| This PR makes a connection pool to help catch a situation where a pre-fork connection pool is used post-fork and generally avoid re-using connections.

This is a Request for Comment and the limitation is that it is only for connection pools and doesn't handle the case where you've already acquired a connection from the pool pre-fork. I think there are other ways to detect this event but figured this was a start of an approach.

### CODE REVIEW

File: lib/pool.py [hunk 1]
```diff
@@ -185,3 +185,24 @@ def closeall(self):
...
+    def __init__(self, *args, **kwargs):
+        self.pid = os.getpid()
+        self.args = args
+        self.kwargs = kwargs
+        super().__init__(*args, **kwargs)
+
+    def getconn(self, *args, **kwargs):
+        current_pid = os.getpid()
+        if current_pid != self.pid:
+            super().__init__(*self.args, **self.kwargs)
```
Comment (dvarrazzo) [inline lib/pool.py:200]: It's not good to just call `__init__` on a class already initialised. I think if you detect this state you should empty the pool: you could add a protected `_flush_pool()` method on the base class to encapsulate it.

File: lib/pool.py [hunk 2]
```diff
@@ -185,3 +185,24 @@ def closeall(self):
             self._closeall()
         finally:
             self._lock.release()
+
+class MultiProcessThreadedConnectionPool(ThreadedConnectionPool):
```
Comment (dvarrazzo) [inline lib/pool.py:189]: Nobody would use a specific subclass if they haven't already gone through some painful debug process as you say. So maybe this should be a change to the base class, not the addition of a new class.

File: lib/pool.py [hunk 3]
```diff
@@ -185,3 +185,24 @@ def closeall(self):
...
+        current_pid = os.getpid()
+        if current_pid != self.pid:
+            super().__init__(*self.args, **self.kwargs)
+            self.pid = current_pid
+        return super().getconn(*args, **kwargs)
+
+    def putconn(self, *args, **kwargs):
+        current_pid = os.getpid()
+        if current_pid != self.pid:
+            raise Exception("Returned a connection to the pool that was acquired pre-fork. Don't use pre-fork connections post-fork.")
```
Comment (dvarrazzo) [inline lib/pool.py:207]: This should probably be a `ProgrammingError`

**Review by dvarrazzo — CHANGES_REQUESTED** [review]

**Comment by dvarrazzo** [comment]
I wonder if an anti-fork protection shouldn't be baked into the connection itself. There is already something partial: forking is detected on delete, see #829, so the connection, internally, has the pid it was created on. If we expose that to python, would you be able to do more robust checking?

**Comment by dvarrazzo** [comment]
Closing as this is a connection problem, not a pool problem.

**Comment by jmo-qap** [comment]
👍  Will circle back on this when I get a chance to check on connection status. I think a connection-based approach would just check if the init pid is different than the current PID and raise an error. Wouldn't want to create a new connection since that would potentially introduce confusing semantics about previous connection state.

**Pull request closed without merging**